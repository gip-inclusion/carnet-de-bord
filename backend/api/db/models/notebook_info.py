from datetime import datetime
from uuid import UUID

from pydantic import BaseModel
from strenum import StrEnum

from api.db.models.orientation_type import OrientationType


class NotebookInfo(BaseModel):
    notebook_id: UUID
    orientation: OrientationType | None
    need_orientation: bool = False
    created_at: datetime
    updated_at: datetime
