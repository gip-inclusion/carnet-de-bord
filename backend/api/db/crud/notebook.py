import json
from typing import Callable
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.account import AccountInfo
from api.db.models.action import Action
from api.db.models.appointment import Appointment
from api.db.models.focus import Focus
from api.db.models.notebook import Notebook
from api.db.models.notebook_member import NotebookMember, NotebookMemberInsert
from api.db.models.rome_code import RomeCode
from api.db.models.target import Target
from api.db.models.wanted_job import WantedJob

NOTEBOOK_BASE_FIELDS = """
n.id as n_id,
n.beneficiary_id as n_beneficiary_id,
n.right_rsa as n_right_rsa,
n.right_rqth as n_right_rqth,
n.right_are as n_right_are,
n.right_ass as n_right_ass,
n.education_level as n_education_level,
n.work_situation_date as n_work_situation_date,
n.right_bonus as n_right_bonus,
n.geographical_area as n_geographical_area,
n.contract_type as n_contract_type,
n.contract_sign_date as n_contract_sign_date,
n.work_situation as n_work_situation,
n.work_situation_end_date as n_work_situation_end_date,
n.contract_start_date as n_contract_start_date,
n.contract_end_date as n_contract_end_date,
n.created_at as n_created_at,
n.updated_at as n_updated_at,
wj.rome_code_id as wj_rome_code_id,
wj.id as wj_id,
rc.code as rc_code,
rc.description as rc_description,
rc.label as rc_label,
nf.id as nf_id,
nf.theme as nf_theme,
nf.situations as nf_situations,
nf.creator_id as nf_creator_id,
nf.notebook_id as nf_notebook_id,
nf.created_at as nf_created_at,
nf.linked_to as nf_linked_to,
nf.updated_at as nf_updated_at,
nt.id as nt_id,
nt.focus_id as nt_focus_id,
nt.target as nt_target,
nt.created_at as nt_created_at,
nt.creator_id as nt_creator_id,
nt.updated_at as nt_updated_at,
nt.status as nt_status,
na.id as na_id,
na.action as na_action,
na.target_id as na_target_id,
na.status as na_status,
na.creator_id as na_creator_id,
na.created_at as na_created_at,
na.updated_at as na_updated_at,
na.initial_id as na_initial_id,
ai_action_creator.firstname as na_firstname,
ai_action_creator.lastname as na_lastname,
ai_action_creator.email as na_email,
nm.id as nm_id,
nm.notebook_id as nm_notebook_id,
nm.account_id as nm_account_id,
nm.last_visited_at as nm_last_visited_at,
nm.member_type as nm_member_type,
nm.last_modified_at as nm_last_modified_at,
nm.created_at as nm_created_at,
nm.creator_id as nm_creator_id,
nm.invitation_sent_at as nm_invitation_sent_at,
nm.active as nm_active,
nap.id as nap_id,
nap.notebook_id as nap_notebook_id,
nap.account_id as nap_account_id,
nap.date as nap_date,
nap.status as nap_status,
nap.created_at as nap_created_at,
nap.updated_at as nap_updated_at,
nap_account_info.firstname as nap_firstname,
nap_account_info.lastname as nap_lastname,
nap_account_info.email as nap_email
"""

NOTEBOOK_BASE_JOINS = """
LEFT JOIN wanted_job wj
ON wj.notebook_id = n.id
LEFT JOIN public.rome_code rc
ON rc.id = wj.rome_code_id
LEFT JOIN notebook_focus nf
ON nf.notebook_id = n.id
LEFT JOIN notebook_target nt
ON nt.focus_id = nf.id
LEFT JOIN notebook_action na
ON na.target_id = nt.id
LEFT JOIN notebook_member nm
ON nm.notebook_id = n.id
LEFT JOIN notebook_appointment nap
ON nap.notebook_id = n.id
LEFT JOIN account_info nap_account_info
ON nap.account_id = nap_account_info.account_id
LEFT JOIN account_info ai_action_creator
ON na.creator_id = ai_action_creator.account_id
"""

NOTEBOOK_BASE_QUERY = (
    "SELECT " + NOTEBOOK_BASE_FIELDS + "FROM public.notebook n" + NOTEBOOK_BASE_JOINS
)


async def add_wanted_jobs_to_notebook(
    record: Record,
    notebook: Notebook,
    notebook_record_prefix: str = "n_",
    wj_record_prefix: str = "wj_",
    rc_record_prefix: str = "rc_",
    notebook_id: str = "notebook_id",
) -> None:
    if record[wj_record_prefix + "rome_code_id"] is not None:
        rome_code = RomeCode(
            id=record[wj_record_prefix + "rome_code_id"],
            code=record[rc_record_prefix + "code"],
            label=record[rc_record_prefix + "label"],
            description=record[rc_record_prefix + "description"],
        )

        wanted_job = WantedJob(
            id=record[wj_record_prefix + "id"],
            notebook_id=record[notebook_record_prefix + notebook_id],
            rome_code_id=record[wj_record_prefix + "rome_code_id"],
            rome_code=rome_code,
        )

        existing_wanted_job: WantedJob | None = next(
            (
                wj
                for wj in notebook.wanted_jobs
                if wj.id == record[wj_record_prefix + "id"]
            ),
            None,
        )

        if not existing_wanted_job:
            notebook.wanted_jobs.append(wanted_job)


async def find_focus(
    notebook: Notebook, find_function: Callable[[Focus], bool]
) -> Focus | None:

    if notebook.focuses is not None:
        return next(
            (f for f in notebook.focuses if find_function(f)),
            None,
        )


async def find_target_from_focus(
    focus: Focus, find_function: Callable[[Target], bool]
) -> Target | None:

    if focus.targets is not None:
        return next(
            (t for t in focus.targets if find_function(t)),
            None,
        )


async def find_target_from_notebook(
    notebook: Notebook, focus_id: UUID, target_id: UUID
) -> Target | None:

    focus: Focus | None = await find_focus(notebook, lambda f: f.id == focus_id)

    if focus is None:
        return

    return await find_target_from_focus(focus, lambda t: t.id == target_id)


async def find_action_from_target(target: Target, action_id: UUID) -> Action | None:

    if target.actions is not None:
        return next(
            (a for a in target.actions if a.id == action_id),
            None,
        )


async def add_action_to_target(
    record: Record,
    notebook: Notebook,
    record_prefix: str = "na_",
    target_record_prefix: str = "nt_",
) -> None:

    target: Target | None = await find_target_from_notebook(
        notebook,
        record[target_record_prefix + "focus_id"],
        record[target_record_prefix + "id"],
    )

    if not target:
        return

    action: Action | None = await find_action_from_target(
        target, record[record_prefix + "id"]
    )

    if action:
        return

    if target.actions is None:
        target.actions = []

    target.actions.append(
        Action(
            id=record[record_prefix + "id"],
            target_id=record[record_prefix + "target_id"],
            action=record[record_prefix + "action"],
            status=record[record_prefix + "status"],
            creator_id=record[record_prefix + "creator_id"],
            created_at=record[record_prefix + "created_at"],
            updated_at=record[record_prefix + "updated_at"],
            initial_id=record[record_prefix + "initial_id"],
            account_info=AccountInfo(
                account_id=record[record_prefix + "creator_id"],
                firstname=record[record_prefix + "firstname"],
                lastname=record[record_prefix + "lastname"],
                email=record[record_prefix + "email"],
            ),
        )
    )


async def add_target_to_focus(
    record: Record,
    notebook: Notebook,
    record_prefix: str = "nt_",
) -> None:

    if record[record_prefix + "id"] is not None:
        focus: Focus | None = await find_focus(
            notebook, lambda f: f.id == record[record_prefix + "focus_id"]
        )

        if focus:
            if focus.targets is None:
                focus.targets = []
            else:
                # Let's see if the target we want to add is already inserted
                target: Target | None = await find_target_from_focus(
                    focus, lambda t: t.id == record[record_prefix + "id"]
                )

                if target:
                    return
                else:
                    focus.targets.append(
                        Target(
                            id=record[record_prefix + "id"],
                            focus_id=record[record_prefix + "focus_id"],
                            target=record[record_prefix + "target"],
                            creator_id=record[record_prefix + "creator_id"],
                            created_at=record[record_prefix + "created_at"],
                            updated_at=record[record_prefix + "updated_at"],
                            status=record[record_prefix + "status"],
                        )
                    )


async def add_focus_to_notebook(
    record: Record,
    notebook: Notebook,
    record_prefix: str = "nf_",
) -> None:
    if record[record_prefix + "id"] is not None:
        if notebook.focuses is None:
            notebook.focuses = []

        focus: Focus | None = next(
            (f for f in notebook.focuses if f.id == record[record_prefix + "id"]), None
        )

        # The current focus is not already attached to the notebook
        # so let's add it
        if not focus:
            notebook.focuses.append(
                Focus(
                    id=record[record_prefix + "id"],
                    theme=record[record_prefix + "theme"],
                    situations=json.loads(record[record_prefix + "situations"]),
                    creator_id=record[record_prefix + "creator_id"],
                    notebook_id=record[record_prefix + "notebook_id"],
                    created_at=record[record_prefix + "created_at"],
                    updated_at=record[record_prefix + "updated_at"],
                    linked_to=record[record_prefix + "linked_to"],
                )
            )


async def add_members_to_notebook(
    record: Record,
    notebook: Notebook,
    record_prefix: str = "nm_",
) -> None:
    if record[record_prefix + "id"] is not None:
        if notebook.members is None:
            notebook.members = []

        notebook_member: NotebookMember | None = next(
            (f for f in notebook.members if f.id == record[record_prefix + "id"]), None
        )

        # The current notebook_member is not already attached to the notebook
        # so let's add it
        if not notebook_member:

            notebook.members.append(
                NotebookMember(
                    id=record[record_prefix + "id"],
                    notebook_id=record[record_prefix + "notebook_id"],
                    account_id=record[record_prefix + "account_id"],
                    last_visited_at=record[record_prefix + "last_visited_at"],
                    member_type=record[record_prefix + "member_type"],
                    last_modified_at=record[record_prefix + "last_modified_at"],
                    invitation_sent_at=record[record_prefix + "invitation_sent_at"],
                    created_at=record[record_prefix + "created_at"],
                    creator_id=record[record_prefix + "creator_id"],
                    active=record[record_prefix + "active"],
                )
            )


async def add_appointments_to_notebook(
    record: Record,
    notebook: Notebook,
    record_prefix: str = "nap_",
) -> None:
    if record[record_prefix + "id"] is not None:
        if notebook.appointments is None:
            notebook.appointments = []

        appointment: Appointment | None = next(
            (f for f in notebook.appointments if f.id == record[record_prefix + "id"]),
            None,
        )

        # The current appointment is not already attached to the notebook
        # so let's add it
        if not appointment:

            notebook.appointments.append(
                Appointment(
                    id=record[record_prefix + "id"],
                    notebook_id=record[record_prefix + "notebook_id"],
                    account_info=AccountInfo(
                        account_id=record[record_prefix + "account_id"],
                        firstname=record[record_prefix + "firstname"],
                        lastname=record[record_prefix + "lastname"],
                        email=record[record_prefix + "email"],
                    ),
                    date=record[record_prefix + "date"],
                    status=record[record_prefix + "status"],
                    created_at=record[record_prefix + "created_at"],
                    updated_at=record[record_prefix + "created_at"],
                )
            )


async def parse_notebook_member_from_record(record: Record) -> NotebookMember:
    return NotebookMember.parse_obj(record)


async def parse_notebooks_from_records(
    records: list[Record],
    notebooks: list[Notebook],
    id_field: str = "n_id",
) -> list[Notebook]:

    for record in records:
        notebook: Notebook | None = next(
            (n for n in notebooks if n.id == record[id_field]), None
        )
        notebook_updated: Notebook | None = await parse_notebook_from_record(
            record, notebook=notebook
        )

        if not notebook:
            notebooks.append(notebook_updated)

    return notebooks


async def parse_notebook_from_record(
    record: Record,
    id_field: str = "id",
    record_prefix: str = "n_",
    beneficiary_id_field: str = "beneficiary_id",
    add_wanted_jobs: bool = True,
    add_focuses: bool = True,
    add_targets: bool = True,
    add_actions: bool = True,
    add_members: bool = True,
    add_appointments: bool = True,
    notebook: Notebook | None = None,
) -> Notebook:

    if not notebook:
        notebook = Notebook(
            id=record[record_prefix + id_field],
            created_at=record[record_prefix + "created_at"],
            updated_at=record[record_prefix + "updated_at"],
            right_rsa=record[record_prefix + "right_rsa"],
            right_rqth=record[record_prefix + "right_rqth"],
            right_are=record[record_prefix + "right_are"],
            right_ass=record[record_prefix + "right_ass"],
            right_bonus=record[record_prefix + "right_bonus"],
            beneficiary_id=record[record_prefix + beneficiary_id_field],
            wanted_jobs=[],
            geographical_area=record[record_prefix + "geographical_area"],
            education_level=record[record_prefix + "education_level"],
            work_situation_date=record[record_prefix + "work_situation_date"],
            contract_type=record[record_prefix + "contract_type"],
            contract_sign_date=record[record_prefix + "contract_sign_date"],
            work_situation=record[record_prefix + "work_situation"],
            work_situation_end_date=record[record_prefix + "work_situation_end_date"],
            contract_start_date=record[record_prefix + "contract_start_date"],
            contract_end_date=record[record_prefix + "contract_end_date"],
        )

    if add_wanted_jobs:
        await add_wanted_jobs_to_notebook(record, notebook, notebook_id=id_field)

    if add_focuses:
        await add_focus_to_notebook(record, notebook)

    if add_targets:
        await add_target_to_focus(record, notebook)

    if add_actions:
        await add_action_to_target(record, notebook)

    if add_members:
        await add_members_to_notebook(record, notebook)

    if add_appointments:
        await add_appointments_to_notebook(record, notebook)

    return notebook


async def get_notebook_member_with_query(
    connection: Connection, query: str, *args
) -> NotebookMember | None:
    async with connection.transaction():

        notebook_member_record: Record | None = await connection.fetchrow(
            "SELECT nm.* FROM public.notebook_member nm " + query,
            *args,
        )

        if notebook_member_record:
            return await parse_notebook_member_from_record(notebook_member_record)


async def get_notebook_members_with_query(
    connection: Connection, query: str, *args
) -> list[NotebookMember]:
    async with connection.transaction():

        records: list[Record] = await connection.fetch(
            "SELECT nm.* FROM public.notebook_member nm " + query,
            *args,
        )

        return [await parse_notebook_member_from_record(record) for record in records]


async def insert_notebook_member(
    connection: Connection, notebook_member_insert: NotebookMemberInsert
) -> NotebookMember | None:

    record = await connection.fetchrow(
        """
        INSERT INTO public.notebook_member (notebook_id, account_id, last_visited_at, member_type, last_modified_at, creator_id, invitation_sent_at, active)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8) returning id, notebook_id, account_id, last_visited_at, member_type, last_modified_at, created_at, creator_id, invitation_sent_at, active
        """,
        *notebook_member_insert.dict().values(),
    )

    if record:
        return await parse_notebook_member_from_record(record)


async def get_notebook_members_by_notebook_id(
    connection: Connection, notebook_id: UUID
) -> list[NotebookMember]:
    return await get_notebook_members_with_query(
        connection,
        "WHERE nm.notebook_id = $1",
        notebook_id,
    )


async def get_notebook_member_by_notebook_id_and_account_id(
    connection: Connection, notebook_id: UUID, account_id: UUID
) -> NotebookMember | None:
    return await get_notebook_member_with_query(
        connection,
        """WHERE nm.notebook_id = $1
        AND nm.account_id = $2""",
        notebook_id,
        account_id,
    )


async def get_notebooks_with_query(
    connection: Connection, query: str, *args
) -> list[Notebook]:
    async with connection.transaction():

        records: list[Record] = await connection.fetch(
            NOTEBOOK_BASE_QUERY + query,
            *args,
        )

        notebooks: list[Notebook] = []
        return await parse_notebooks_from_records(records=records, notebooks=notebooks)


async def get_notebook_with_query(
    connection: Connection, query: str, *args
) -> Notebook | None:
    async with connection.transaction():

        record: Record = await connection.fetchrow(
            NOTEBOOK_BASE_QUERY + query,
            *args,
        )

        return await parse_notebook_from_record(record)


async def get_notebooks_by_structure_id(
    connection: Connection, structure_id: UUID
) -> list[Notebook]:
    return await get_notebooks_with_query(
        connection,
        """LEFT JOIN account ON nm.account_id = account.id
        LEFT JOIN professional ON account.professional_id = professional.id
        WHERE professional.structure_id = $1""",
        structure_id,
    )


async def get_notebook_by_id(
    connection: Connection, notebook_id: UUID
) -> Notebook | None:
    return await get_notebook_with_query(
        connection,
        """WHERE n.id = $1""",
        notebook_id,
    )
