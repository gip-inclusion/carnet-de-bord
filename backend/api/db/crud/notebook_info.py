from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.notebook_info import NotebookInfo, OrientationType


def parse_notebook_info(record: Record) -> NotebookInfo:
    return NotebookInfo.parse_obj(record)


async def insert_or_update_need_orientation(
    connection: Connection,
    notebook_id: UUID,
    orientation: OrientationType | None,
    need_orientation: bool,
) -> NotebookInfo | None:
    record = await connection.fetchrow(
        """
        INSERT INTO public.notebook_info (notebook_id, orientation, need_orientation)
        VALUES ($1, $2, $3)
        ON CONFLICT ON CONSTRAINT notebook_info_pkey
        DO UPDATE SET need_orientation = $3
        RETURNING notebook_id, orientation, need_orientation, created_at, updated_at
        """,
        notebook_id,
        orientation,
        need_orientation,
    )

    if record:
        return parse_notebook_info(record)


async def get_notebook_info(
    connection: Connection, notebook_id: UUID
) -> NotebookInfo | None:
    record = await connection.fetchrow(
        """
        SELECT * from public.notebook_info WHERE notebook=$1
        """,
        notebook_id,
    )

    if record:
        return NotebookInfo.parse_obj(record)
