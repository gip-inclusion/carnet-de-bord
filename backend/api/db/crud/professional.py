import logging

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.professional import Professional, ProfessionalInsert

logger = logging.getLogger(__name__)


async def get_professional_with_query(
    connection: Connection, query: str, *args
) -> Professional | None:
    async with connection.transaction():

        professional_record: Record | None = await connection.fetchrow(
            "SELECT public.professional.*, public.account.id as account_id "
            "FROM public.professional "
            "LEFT JOIN public.account "
            "ON public.account.professional_id = public.professional.id " + query,
            *args,
        )

        if professional_record:
            return Professional.parse_obj(professional_record)


async def get_professional_by_email(
    connection: Connection, email: str
) -> Professional | None:
    return await get_professional_with_query(
        connection,
        "WHERE LOWER(public.professional.email) = LOWER($1)",
        email,
    )


async def insert_professional(
    connection: Connection, professional_insert: ProfessionalInsert
) -> Professional | None:

    record = await connection.fetchrow(
        """
        INSERT INTO public.professional (structure_id, email, lastname, firstname, position, mobile_number)
        VALUES ($1, $2, $3, $4, $5, $6) returning id, structure_id, email, lastname, firstname, position, mobile_number, created_at, updated_at
        """,
        *professional_insert.dict().values(),
    )

    if record:
        return await parse_professional_from_record(record)


async def parse_professional_from_record(record: Record) -> Professional:
    return Professional.parse_obj(record)
