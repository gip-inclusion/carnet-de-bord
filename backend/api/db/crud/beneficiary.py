from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from asyncpg.exceptions import UniqueViolationError

from api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    parse_notebook_from_record,
)
from api.db.models.beneficiary import Beneficiary, BeneficiaryImport

BENEFICIARY_BASE_QUERY = (
    """SELECT b.*, acc.id as account_id,"""
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def select_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    return await connection.fetch(
        f"""
SELECT (firstname, lastname, date_of_birth, internal_id, deployment_id, id)
FROM beneficiary
WHERE (firstname = $1 AND lastname = $2 AND date_of_birth = $3)
OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.si_id,
        deployment_id,
    )


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
UPDATE beneficiary SET mobile_number = $1
where internal_id = $2
returning id
        """,
        beneficiary.phone_number,
        beneficiary.si_id,
    )


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
INSERT INTO BENEFICIARY (firstname, lastname, internal_id, date_of_birth, deployment_id, mobile_number)
values ($1, $2, $3, $4, $5, $6)
returning id
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.si_id,
        beneficiary.date_of_birth,
        deployment_id,
        beneficiary.phone_number,
    )


async def import_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    result = select_beneficiary(connection, beneficiary, deployment_id)
    existing_rows = await result
    if len(existing_rows) == 0:
        print("inserting new user")
        await insert_beneficiary(connection, beneficiary, deployment_id)
    else:
        print(existing_rows[0][0][4])
        if (
            len(existing_rows) == 1
            and existing_rows[0][0][0] == beneficiary.firstname
            and existing_rows[0][0][1] == beneficiary.lastname
            and existing_rows[0][0][2] == beneficiary.date_of_birth
            and existing_rows[0][0][3] == beneficiary.si_id
        ):
            # Doesn't work somehow
            #     and existing_rows[0][0][4] == deployment_id
            print("updating existing user")
            await update_beneficiary(connection, beneficiary, deployment_id)
        else:
            print("data problem")


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:

    async with connection.transaction():

        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:

            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)

            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )
