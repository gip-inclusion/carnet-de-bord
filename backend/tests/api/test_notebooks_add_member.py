from unittest import mock

import pytest
from asyncpg.connection import Connection
from httpx import AsyncClient
from syrupy.data import Snapshot

from cdb.api.db.crud.beneficiary import get_structures_for_beneficiary
from cdb.api.db.crud.notebook import get_notebook_members_by_notebook_id
from cdb.api.db.crud.notebook_info import get_notebook_info
from cdb.api.db.models.notebook import Notebook
from cdb.api.db.models.professional import Professional
from tests.utils.assert_helpers import assert_member, assert_structure

pytestmark = pytest.mark.graphql


@mock.patch("cdb.api.core.emails.send_mail")
async def test_verify_no_token(
    _: mock.Mock,
    test_client: AsyncClient,
    notebook_sophie_tifour: Notebook,
):
    response = await test_client.post(
        f"/v1/notebooks/{notebook_sophie_tifour.id}/members",
    )
    assert response.status_code == 401
    json = response.json()
    assert json["detail"] == "Missing credentials"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_orientation_manager_not_allowed_to_add_notebook_member(
    _: mock.Mock,
    test_client: AsyncClient,
    giulia_diaby_jwt: str,
    notebook_sophie_tifour: Notebook,
):
    response = await test_client.post(
        f"/v1/notebooks/{notebook_sophie_tifour.id}/members",
        headers={"Authorization": "Bearer " + giulia_diaby_jwt},
    )
    assert response.status_code == 403
    json = response.json()
    assert json["detail"] == "Operation forbidden to the given role"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_add_notebook_member_as_no_referent_with_no_structure_id_in_token(
    _: mock.Mock,
    test_client: AsyncClient,
    notebook_sophie_tifour: Notebook,
    get_professional_jwt_without_structure_id: str,
):
    response = await test_client.post(
        f"/v1/notebooks/{str(notebook_sophie_tifour.id)}/members",
        json={"member_type": "no_referent"},
        headers={
            "Authorization": "Bearer " + f"{get_professional_jwt_without_structure_id}"
        },
    )
    assert response.status_code == 403
    json = response.json()
    assert json["detail"] == "Unsufficient permission (structureId is missing)"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_add_notebook_member_as_no_referent(
    mock_send_email: mock.Mock,
    test_client: AsyncClient,
    notebook_sophie_tifour: Notebook,
    get_professional_paul_camara_jwt: str,
    professional_paul_camara: Professional,
    professional_pierre_chevalier: Professional,
    db_connection: Connection,
):
    former_notebook_info = await get_notebook_info(
        db_connection,
        notebook_sophie_tifour.id,
    )
    response = await test_client.post(
        f"/v1/notebooks/{notebook_sophie_tifour.id}/members",
        json={"member_type": "no_referent"},
        headers={"Authorization": "Bearer " + get_professional_paul_camara_jwt},
    )
    assert response.status_code == 204
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_sophie_tifour.id,
    )
    structures = await get_structures_for_beneficiary(
        db_connection,
        notebook_sophie_tifour.beneficiary_id,
    )
    # Check that a new member was added
    assert_member(members, professional_paul_camara, "no_referent", True)

    # Check that beneficiary_structure has not changed
    assert_structure(
        structures,
        beneficiary_status="current",
        structure_name="Centre Communal d'action social Livry-Gargan",
    )
    mock_send_email.assert_not_called()
    # Check that former referent is still the referent
    assert_member(members, professional_pierre_chevalier, "referent", True)
    notebook_info = await get_notebook_info(
        db_connection,
        notebook_sophie_tifour.id,
    )
    # Check that orientation reason has not been changed
    assert notebook_info.orientation_reason == former_notebook_info.orientation_reason


@mock.patch("cdb.api.core.emails.send_mail")
async def test_add_notebook_member_as_referent(
    mock_send_email: mock.Mock,
    snapshot: Snapshot,
    test_client: AsyncClient,
    notebook_sophie_tifour: Notebook,
    get_professional_paul_camara_jwt: str,
    professional_paul_camara: Professional,
    professional_pierre_chevalier: Professional,
    db_connection: Connection,
):
    [orientation_system_id] = await db_connection.fetchrow(
        """
        SELECT orientation_system_id
        FROM professional_orientation_system
        WHERE professional_id = $1
        """,
        professional_paul_camara.id,
    )
    response = await test_client.post(
        f"/v1/notebooks/{notebook_sophie_tifour.id}/members",
        json={"member_type": "referent", "orientation": str(orientation_system_id)},
        headers={"Authorization": "Bearer " + get_professional_paul_camara_jwt},
    )
    assert response.status_code == 204
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_sophie_tifour.id,
    )

    # Check that a new member was added
    assert_member(members, professional_paul_camara, "referent", True)
    # Check that former referent has been deactivated
    assert_member(members, professional_pierre_chevalier, "referent", False)
    # Check that former referent is still an active member
    assert_member(members, professional_pierre_chevalier, "no_referent", True)
    # Check that an email is sent to former referent
    email_former_referent = mock_send_email.call_args_list[0]
    assert snapshot == email_former_referent

    structures = await get_structures_for_beneficiary(
        db_connection,
        notebook_sophie_tifour.beneficiary_id,
    )

    # Check that former structure has been set as 'outdated'
    assert_structure(
        structures,
        beneficiary_status="outdated",
        structure_name="Centre Communal d'action social Livry-Gargan",
    )

    # Check that new structure has been added as 'current'
    assert_structure(
        structures,
        beneficiary_status="current",
        structure_name="Service Social Départemental",
    )

    notebook_info = await get_notebook_info(
        db_connection,
        notebook_sophie_tifour.id,
    )
    assert notebook_info.need_orientation is False
    assert notebook_info.orientation_system_id == orientation_system_id
    assert notebook_info.orientation_reason is None


@mock.patch("cdb.api.core.emails.send_mail")
async def test_add_notebook_member_as_referent_orientation_not_available(
    mock_send_email: mock.Mock,
    test_client: AsyncClient,
    notebook_sophie_tifour: Notebook,
    get_professional_paul_camara_jwt: str,
    professional_paul_camara: Professional,
    professional_pierre_chevalier: Professional,
    db_connection: Connection,
):
    [orientation_system_id] = await db_connection.fetchrow(
        """
        SELECT orientation_system_id
        FROM professional_orientation_system
        WHERE professional_id = $1
        """,
        professional_paul_camara.id,
    )
    await db_connection.execute(
        """
        DELETE FROM professional_orientation_system
        WHERE orientation_system_id = $1
        """,
        orientation_system_id,
    )
    response = await test_client.post(
        f"/v1/notebooks/{notebook_sophie_tifour.id}/members",
        json={
            "member_type": "referent",
            "orientation": str(orientation_system_id),
        },
        headers={"Authorization": "Bearer " + get_professional_paul_camara_jwt},
    )
    assert response.status_code == 422
    assert response.json() == {
        "detail": [
            "Le référent doit être rattaché au dispositif d’accompagnement spécifié."
        ],
        "message": [
            "Le référent doit être rattaché au dispositif d’accompagnement spécifié."
        ],
    }
    members = await get_notebook_members_by_notebook_id(
        db_connection,
        notebook_sophie_tifour.id,
    )

    assert all(
        member.account_id != professional_paul_camara.account_id for member in members
    )
    assert_member(members, professional_pierre_chevalier, "referent", active=True)
    mock_send_email.assert_not_called()

    structures = await get_structures_for_beneficiary(
        db_connection,
        notebook_sophie_tifour.beneficiary_id,
    )
    assert_structure(
        structures,
        beneficiary_status="current",
        structure_name="Centre Communal d'action social Livry-Gargan",
    )

    notebook_info = await get_notebook_info(db_connection, notebook_sophie_tifour.id)
    # Ensure orientation system has not been updated with not available one
    assert notebook_info.orientation_system_id != orientation_system_id
