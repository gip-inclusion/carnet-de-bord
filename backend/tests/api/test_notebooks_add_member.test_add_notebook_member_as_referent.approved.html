


<!DOCTYPE html>
<html lang="fr">
	<head>
		<meta charset="UTF-8" />
		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>Réorientation d’un bénéficiaire</title>
	</head>
	<body>
		<h1>
			Réorientation d’un bénéficiaire
		</h1>
		<p>Bénéficiaire&nbsp;: <strong>Sophie Tifour</strong></p>

		
		
		<div style="display: flex">
			<p>Suivi précédent&nbsp;:</p>
			<div>
				<ul style="list-style-type:none; margin: 0; padding: 0">
					
					<li><strong>Pierre Chevalier (Centre Communal d&#39;action social Livry-Gargan)</strong></li>
					
				</ul>
			</div>
		</div>
		<p>
			Nouveau suivi&nbsp;:<strong>Paul Camara (Service Social Départemental)</strong>
		</p>
		

		<p>
			Pour accéder à Carnet de bord, veuillez cliquer sur le lien
			ci-dessous&nbsp;:
		</p>
		<p style="padding-left: 20%">
			<a
				rel="nofollow"
				href="http://localhost:3001"
				style="
					background-color: #6a20ae;
					font-size: 14px;
					font-family: Helvetica, Arial, sans-serif;
					font-weight: bold;
					text-decoration: none;
					padding: 8px 10px;
					color: #ffffff;
					border-radius: 5px;
					display: inline-block;
				"
			>
				<span>Accédez à Carnet de bord</span>
			</a>
		</p>
		<p>L'équipe Carnet de bord</p>
	</body>
</html>
